### PR TITLE
Add flir lepton support to the fir module

### PR DESCRIPTION
**Pull request by kwagyeman**
Adds support for an external flir lepton module to all OpenMV Cams.

This will allow SignTown to make a flir lepton shield for cameras.

Also, this give the OpenMV Pure Thermal support for the FLIR Lepton.

...

I'll send a PR next for automatic support for detecting the attached sensor and then update the examples with this to avoid copying the thermal examples again.

### CODE REVIEW

File: src/omv/ports/stm32/modules/py_fir_lepton.c [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+    fir_adc_resolution = fir_lepton_rad_en ? 16 : 14;
+
+#if defined(OMV_FIR_LEPTON_VSYNC_PRESENT)
+    if (LEP_SetOemGpioMode(&fir_lepton_handle, LEP_OEM_GPIO_MODE_VSYNC) != LEP_OK) {
+        return -9;
+    }
+
+    extint_register((mp_obj_t) OMV_FIR_LEPTON_VSYNC_PIN, GPIO_MODE_IT_FALLING, GPIO_PULLUP,
+                    (mp_obj_t) &fir_lepton_extint_callback_obj, true);
+#endif
```
Comment (iabdalkader) [inline src/omv/ports/stm32/modules/py_fir_lepton.c:438]:
```suggestion
    #if defined(OMV_FIR_LEPTON_VSYNC_PRESENT)
    if (LEP_SetOemGpioMode(&fir_lepton_handle, LEP_OEM_GPIO_MODE_VSYNC) != LEP_OK) {
        return -9;
    }

    extint_register((mp_obj_t) OMV_FIR_LEPTON_VSYNC_PIN, GPIO_MODE_IT_FALLING, GPIO_PULLUP,
                    (mp_obj_t) &fir_lepton_extint_callback_obj, true);
    #endif
```

File: src/omv/ports/stm32/modules/py_fir_lepton.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+void fir_lepton_deinit();
+int fir_lepton_init();
+void fir_lepton_register_vsync_cb(mp_obj_t cb);
+mp_obj_t fir_lepton_get_radiometry();
+mp_obj_t fir_lepton_read_ta();
+mp_obj_t fir_lepton_read_ir(bool mirror, bool flip, bool transpose);
+void fir_lepton_fill_image(image_t *img, bool auto_range, float min, float max,
+                           bool mirror, bool flip, bool transpose);
+void fir_lepton_trigger_ffc();
+void fir_lepton_wait_on_ffc(uint n_args, const mp_obj_t *args);
```
Comment (iabdalkader) [inline src/omv/ports/stm32/modules/py_fir_lepton.h:22]:
```suggestion
```

File: src/omv/ports/stm32/modules/py_fir_lepton.c [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+            break;
+        }
+
+        if ((systick_current_millis() - start) >= delay_ms) {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "Timeout!"));
+        }
+
+        systick_sleep(1);
+    }
+}
```
Comment (iabdalkader) [inline src/omv/ports/stm32/modules/py_fir_lepton.c:693]:
```suggestion
void fir_lepton_trigger_ffc()
{
    if (LEP_RunSysFFCNormalization(&fir_lepton_handle) != LEP_OK) {
        nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "FFC Error!"));
    }
}

void fir_lepton_wait_on_ffc(uint n_args, const mp_obj_t *args)
{
    uint32_t start = systick_current_millis();
    uint32_t delay_ms = (n_args < 1) ? 5000 : mp_obj_get_int(args[0]);

    for (;;) {
        LEP_SYS_STATUS_E status;

        if (LEP_GetSysFFCStatus(&fir_lepton_handle, &status) != LEP_OK) {
            nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "SYS Error!"));
        }

        if (status == LEP_SYS_STATUS_READY) {
            break;
        }

        if ((systick_current_millis() - start) >= delay_ms) {
            nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "Timeout!"));
        }

        systick_sleep(1);
    }
}
```
merge in 1 function.

File: src/omv/modules/py_fir.c [hunk 4]
```diff
@@ -228,6 +233,12 @@ static mp_obj_t fir_get_ir(int w, int h, float Ta, float *To, bool mirror, bool
 
 static mp_obj_t py_fir_deinit()
 {
+#if defined(OMV_FIR_LEPTON_PRESENT)
+    if (fir_sensor == FIR_LEPTON) {
+        fir_lepton_deinit();
+    }
+#endif
+
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:241]:
```suggestion
    #if defined(OMV_FIR_LEPTON_PRESENT)
    if (fir_sensor == FIR_LEPTON) {
        fir_lepton_deinit();
    }
    #endif

```

File: src/omv/modules/py_fir.c [hunk 5]
```diff
@@ -52,20 +54,23 @@
...
 
 static enum {
     FIR_NONE,
     FIR_MLX90621,
     FIR_MLX90640,
-    FIR_AMG8833
+    FIR_AMG8833,
+#if defined(OMV_FIR_LEPTON_PRESENT)
+    FIR_LEPTON
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:67]:
```suggestion
    #if defined(OMV_FIR_LEPTON_PRESENT)
    FIR_LEPTON
    #endif
```

File: src/omv/modules/py_fir.c [hunk 6]
```diff
@@ -52,20 +54,23 @@
...
 } fir_sensor = FIR_NONE;
 
-static int fir_width = 0;
-static int fir_height = 0;
-static int fir_ir_fresh_rate = 0;
-static int fir_adc_resolution = 0;
+int fir_width = 0;
+int fir_height = 0;
+int fir_ir_fresh_rate = 0;
+int fir_adc_resolution = 0;
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:73]: Any way to keep these static ? Maybe put them in a struct and pass a pointer around ?

File: src/omv/modules/py_fir.c [hunk 7]
```diff
@@ -369,6 +380,29 @@ mp_obj_t py_fir_init(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
...
+                } else {
+                    py_fir_deinit();
+                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError,
+                                                       "Failed to init the Lepton!"));
+                }
+            }
+
+            return mp_const_none;
+        }
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:405]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON: {
            fir_sensor = FIR_LEPTON;
            FIR_LEPTON_RETRY:
            cambus_init(&fir_bus, OMV_FIR_LEPTON_I2C_BUS, OMV_FIR_LEPTON_I2C_BUS_SPEED);

            int error = fir_lepton_init();

            if (error != 0) {
                if (first_init) {
                    first_init = false;
                    cambus_pulse_scl(&fir_bus);
                    goto FIR_LEPTON_RETRY;
                } else {
                    py_fir_deinit();
                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError,
                                                       "Failed to init the Lepton!"));
                }
            }

            return mp_const_none;
        }
        #endif
```

File: src/omv/modules/py_fir.c [hunk 8]
```diff
@@ -412,6 +446,10 @@ static mp_obj_t py_fir_refresh()
             return mp_obj_new_int(mlx_90640_refresh_rates[fir_ir_fresh_rate]);
         case FIR_AMG8833:
             return mp_obj_new_int(fir_ir_fresh_rate);
+#if defined(OMV_FIR_LEPTON_PRESENT)
+        case FIR_LEPTON:
+            return mp_obj_new_int(fir_ir_fresh_rate);
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:452]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON:
            return mp_obj_new_int(fir_ir_fresh_rate);
        #endif
```

File: src/omv/modules/py_fir.c [hunk 9]
```diff
@@ -429,12 +467,60 @@ static mp_obj_t py_fir_resolution()
             return mp_obj_new_int(fir_adc_resolution + 16);
         case FIR_AMG8833:
             return mp_obj_new_int(fir_adc_resolution);
+#if defined(OMV_FIR_LEPTON_PRESENT)
+        case FIR_LEPTON:
+            return mp_obj_new_int(fir_adc_resolution);
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:473]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON:
            return mp_obj_new_int(fir_adc_resolution);
        #endif
```

File: src/omv/modules/py_fir.c [hunk 10]
```diff
@@ -467,6 +553,11 @@ mp_obj_t py_fir_read_ta()
             PY_ASSERT_TRUE_MSG((error == 0), "Failed to read the AMG8833 sensor data!");
             return mp_obj_new_float(AMG8833_12_TO_16(temp) * 0.0625f);
         }
+#if defined(OMV_FIR_LEPTON_PRESENT)
+        case FIR_LEPTON: {
+            return fir_lepton_read_ta();
+        }
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:560]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON: {
            return fir_lepton_read_ta();
        }
        #endif
```

File: src/omv/modules/py_fir.c [hunk 11]
```diff
@@ -510,6 +601,11 @@ mp_obj_t py_fir_read_ir(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
             fb_alloc_free_till_mark();
             return result;
         }
+#if defined(OMV_FIR_LEPTON_PRESENT)
+        case FIR_LEPTON: {
+            return fir_lepton_read_ir(arg_hmirror, arg_vflip, fir_transposed);
+        }
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:608]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON: {
            return fir_lepton_read_ir(arg_hmirror, arg_vflip, fir_transposed);
        }
        #endif
```

File: src/omv/modules/py_fir.c [hunk 12]
```diff
@@ -831,6 +927,13 @@ mp_obj_t py_fir_snapshot(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
             fb_free();
             break;
         }
+#if defined(OMV_FIR_LEPTON_PRESENT)
+        case FIR_LEPTON: {
+            fir_lepton_fill_image(&src_img, !scale_obj, min, max,
+                                  arg_hmirror, arg_vflip, arg_transpose);
+            break;
+        }
+#endif
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:936]:
```suggestion
        #if defined(OMV_FIR_LEPTON_PRESENT)
        case FIR_LEPTON: {
            fir_lepton_fill_image(&src_img, !scale_obj, min, max,
                                  arg_hmirror, arg_vflip, arg_transpose);
            break;
        }
        #endif
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: src/omv/ports/stm32/modules/py_fir_lepton.c [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+ * This work is licensed under the MIT license, see the file LICENSE for details.
+ *
+ * FIR Python module.
+ */
+#include "py/nlr.h"
+#include "py/runtime.h"
+#include "py/obj.h"
+
+#include "extint.h"
+#include "spi.h"
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply src/omv/ports/stm32/modules/py_fir_lepton.c:16]: This is needed per reusing MP's init methods.

**Pull request merged**